src/libraries/System.Private.CoreLib/src/System/IO/AsyncWindowsFileStreamStrategy.cs

```diff
@@ -0,0 +1,406 @@
+
+        public override int Read(byte[] buffer, int offset, int count)
+            => ReadAsyncInternal(new Memory<byte>(buffer, offset, count)).GetAwaiter().GetResult();
+
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+            => ReadAsyncInternal(new Memory<byte>(buffer, offset, count), cancellationToken);
+
+        public override ValueTask<int> ReadAsync(Memory<byte> destination, CancellationToken cancellationToken = default)
+            => new ValueTask<int>(ReadAsyncInternal(destination, cancellationToken));
+
+        private unsafe Task<int> ReadAsyncInternal(Memory<byte> destination, CancellationToken cancellationToken = default)
+        {
+            if (!CanRead)
+            {
+                throw Error.GetReadNotSupported();
+            }
+
+            Debug.Assert(!_fileHandle.IsClosed, "!_handle.IsClosed");
+
+            // Create and store async stream class library specific data in the async result
+            FileStreamCompletionSource completionSource = FileStreamCompletionSource.Create(this, _preallocatedOverlapped, 0, destination);
+            NativeOverlapped* intOverlapped = completionSource.Overlapped;
+
+            // Calculate position in the file we should be at after the read is done
+            if (CanSeek)
+            {
+                long len = Length;
+
+                // Make sure we are reading from the position that we think we are
+                VerifyOSHandlePosition();
+
+                if (_filePosition + destination.Length > len)
+                {
+                    if (_filePosition <= len)
+                    {
+                        destination = destination.Slice(0, (int)(len - _filePosition));
+                    }
+                    else
+                    {
+                        destination = default;
+                    }
+                }
+
+                // Now set the position to read from in the NativeOverlapped struct
+                // For pipes, we should leave the offset fields set to 0.
+                intOverlapped->OffsetLow = unchecked((int)_filePosition);
+                intOverlapped->OffsetHigh = (int)(_filePosition >> 32);
+
+                // When using overlapped IO, the OS is not supposed to
+                // touch the file pointer location at all.  We will adjust it
+                // ourselves. This isn't threadsafe.
+
+                // WriteFile should not update the file pointer when writing
+                // in overlapped mode, according to MSDN.  But it does update
+                // the file pointer when writing to a UNC path!
+                // So changed the code below to seek to an absolute
+                // location, not a relative one.  ReadFile seems consistent though.
+                SeekCore(_fileHandle, destination.Length, SeekOrigin.Current);
```

Comment:
We'll also be getting rid of this?

Reply:
Yes (https://github.com/dotnet/runtime/pull/49145)! Removing that (https://github.com/dotnet/runtime/commit/91423fa3d672ed622c758b57f570b7a25bc12b33) is going to fix #16354 and moreover allow for https://github.com/dotnet/runtime/commit/a7ca4cb277aae5b3f2c3f15f6b1e0058d0714589 and solving #25905, and give us an optimal sys calls usage and close-to perfect perf (reducing allocaitons will give perfect perf).